### PR TITLE
feat(aws-logs): allow importing `CrossAccountDestination` by ARN

### PR DESCRIPTION
**Pull request by rgrochowicz**
This PR adds `CrossAccountDestination.fromDestinationArn` to support importing `CrossAccountDestination` from an ARN. This allows for a better cross-account experience with CloudWatch Logs when one wants to send logs from one account to another. Without the ability to import by ARN, users must either use the Cfn resources directly or implement their own `ILogSubscriptionDestination`.

Usually, one creates a CloudWatch Logs destination in one account, gives it a destination policy, and collects its ARN. Then, one uses that ARN in a second account as a target for a `SubscriptionFilter` to send logs back to the first account. For more information on the cross-account flow, see [Cross-account log data sharing with subscriptions](https://docs.aws.amazon.com/AmazonCloudWatch/latest/logs/CrossAccountSubscriptions.html).

Related issues:  #11154 #18600 #9852
Closes: #6884 #15451

----

### All Submissions:

* [x] Have you followed the guidelines in our [Contributing guide?](https://github.com/aws/aws-cdk/blob/master/CONTRIBUTING.md)

### Adding new Unconventional Dependencies:

* [ ] This PR adds new unconventional dependencies following the process described [here](https://github.com/aws/aws-cdk/blob/master/CONTRIBUTING.md/#adding-new-unconventional-dependencies)

### New Features

* [ ] Have you added the new feature to an [integration test](https://github.com/aws/aws-cdk/blob/master/INTEGRATION_TESTS.md)?
	* [ ] Did you use `yarn integ` to deploy the infrastructure and generate the snapshot (i.e. `yarn integ` without `--dry-run`)?

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-logs/README.md [hunk 1]
```diff
@@ -119,6 +114,22 @@ new logs.SubscriptionFilter(this, 'Subscription', {
 });
 ```
 
+If your `SubscriptionFilter` destination lives in a different account, a `CrossAccountDestination`
+object needs to be added in the destination account which will act as a proxy
+for the remote resource. After you have created the `CrossAccountDestination` in the destination account,
+use the `CrossAccountDestination.fromDestinationArn` method in the source account to import the destination:
```
Comment (comcalvi) [inline packages/@aws-cdk/aws-logs/README.md:120]:
```
This object is automatically created for you
if you use the CDK Kinesis library.
```

This text has been deleted, but these changes don't indicate that the behavior will. It's worth mentioning a list of resources which will automatically create the destination for you, even if it's just Kinesis.

File: packages/@aws-cdk/aws-logs/lib/cross-account-destination.ts [hunk 2]
```diff
@@ -43,7 +61,26 @@ export interface CrossAccountDestinationProps {
  *
  * @resource AWS::Logs::Destination
  */
-export class CrossAccountDestination extends cdk.Resource implements ILogSubscriptionDestination {
+export class CrossAccountDestination extends CrossAccountDestinationBase {
+  /**
+   * Import an existing CloudWatch Logs Destination given its ARN.
+   *
+   * @param scope The parent creating construct (usually `this`).
+   * @param id The construct's name.
```
Comment (comcalvi) [inline packages/@aws-cdk/aws-logs/lib/cross-account-destination.ts:69]:
```suggestion
   * @param id construct id
```

File: packages/@aws-cdk/aws-logs/lib/cross-account-destination.ts [hunk 3]
```diff
@@ -6,6 +6,24 @@ import { CfnDestination } from './logs.generated';
 import { ILogSubscriptionDestination, LogSubscriptionDestinationConfig } from './subscription-filter';
 import { ArnFormat } from '@aws-cdk/core';
 
+abstract class CrossAccountDestinationBase extends cdk.Resource implements ILogSubscriptionDestination {
```
Comment (comcalvi) [inline packages/@aws-cdk/aws-logs/lib/cross-account-destination.ts:9]: Why are we adding this base class? I know this is a common pattern elsewhere, but this base class doesn't add much value...are there multiple other types of `CrossAccountDestination` that would actually inherit from this?

Comment (comcalvi) [inline packages/@aws-cdk/aws-logs/lib/cross-account-destination.ts:9]: Also missing docstring.

File: packages/@aws-cdk/aws-logs/lib/cross-account-destination.ts [hunk 4]
```diff
@@ -43,7 +61,26 @@ export interface CrossAccountDestinationProps {
...
+   *
+   * @param scope The parent creating construct (usually `this`).
+   * @param id The construct's name.
+   * @param destinationArn AWS CloudWatch Logs Destination ARN (i.e. arn:aws:logs:<region>:<account-id>:destination:MyDestination).
+   */
+  public static fromDestinationArn(scope: Construct, id: string, destinationArn: string): ILogSubscriptionDestination {
+    const stack = cdk.Stack.of(scope);
+    const destinationName = stack.splitArn(destinationArn, ArnFormat.COLON_RESOURCE_NAME).resourceName!;
+
+    class Import extends CrossAccountDestinationBase {
```
Comment (comcalvi) [inline packages/@aws-cdk/aws-logs/lib/cross-account-destination.ts:76]: I don't think we need the base class at all, `Import` can just extend `Resource` and implement `ILogSubscriptionDestination` like the `Import` in `fromRoleArn`: https://github.com/comcalvi/aws-cdk/blob/c97810359a3c00758a8e23621be53e9a7d204a7d/packages/%40aws-cdk/aws-iam/lib/role.ts#L218.

File: packages/@aws-cdk/aws-logs/test/destination.test.ts [hunk 5]
```diff
@@ -65,4 +65,24 @@ describe('destination', () => {
       TargetArn: 'arn:bogus',
     });
   });
+
+  test('fromDestinationArn', () => {
+    // GIVEN
+    const stack = new cdk.Stack();
```
Comment (comcalvi) [inline packages/@aws-cdk/aws-logs/test/destination.test.ts:71]: This is a good test case, but I'd like to see two more:

1. Cross stack, within the same account
2. Cross stack, between two different accounts

File: packages/@aws-cdk/aws-logs/test/destination.test.ts [hunk 6]
```diff
@@ -65,4 +65,24 @@ describe('destination', () => {
       TargetArn: 'arn:bogus',
     });
   });
+
+  test('fromDestinationArn', () => {
```
Comment (comcalvi) [inline packages/@aws-cdk/aws-logs/test/destination.test.ts:69]: We also need an integ test that verifies the cross account behavior.

**Review by comcalvi — CHANGES_REQUESTED** [review]
Thanks for the contribution! This is a good first attempt.

File: packages/@aws-cdk/aws-logs/README.md (hunk 1 above)
Comment (rgrochowicz) [reply packages/@aws-cdk/aws-logs/README.md:120]: From what I can find, this behavior doesn't currently exist. It was added in https://github.com/aws/aws-cdk/commit/9d58d7a91a8ad06d7460c408733ac54ebc796a6e#diff-91d106241cb46018eea8934193bfcab27e75c19393ea7472382bccf5c776e8ba and removed in https://github.com/aws/aws-cdk/commit/01601c2c9218da573d0c94f72ea0639551a7160b#diff-04dca47d05b6466108d9f1a1742ddbfc5544fe3221e3efde3741a928d0ee4565.

The link between Kinesis and CloudWatch was shifted to the `@aws-cdk/aws-logs-destinations` package, and these docs weren't updated at the time.

File: packages/@aws-cdk/aws-logs/lib/cross-account-destination.ts (hunk 4 above)
Comment (rgrochowicz) [reply packages/@aws-cdk/aws-logs/lib/cross-account-destination.ts:76]: Yeah, I was torn between the two ways of representing imported resources.

The one benefit of using a `Base` class (and why I went that route) is that the `bind()` functions are equivalent between the imported and created resources. I'll change this back.

File: packages/@aws-cdk/aws-logs/test/destination.test.ts (hunk 6 above)
Comment (rgrochowicz) [reply packages/@aws-cdk/aws-logs/test/destination.test.ts:69]: Would it be ok to pull in the dependencies required to execute an integration test even if those dependencies cause a circular dependency chain?

To properly create a destination, we'll have to create a Lambda function and its associated resources.

Comment (rgrochowicz) [reply packages/@aws-cdk/aws-logs/test/destination.test.ts:69]: Also, do we have an existing integration test that uses multiple accounts? As I understand it, the integration test runner only executes from within the context of one account. How could I setup a multi-account integration test?

Comment (comcalvi) [reply packages/@aws-cdk/aws-logs/test/destination.test.ts:69]: we do indeed support cross account integ tests. Check out this example in `core`: https://github.com/aws/aws-cdk/blob/main/packages/%40aws-cdk/core/test/cross-environment-token.test.ts#L9

Comment (comcalvi) [reply packages/@aws-cdk/aws-logs/test/destination.test.ts:69]: You should not have to create circular dependencies in the integ tests. This likely indicates that the test belongs in the integration module, `aws-logs-destinations`, instead.

**Comment by comcalvi** [comment]
Please see the integ test guide if you're unsure how to proceed: https://github.com/aws/aws-cdk/blob/main/INTEGRATION_TESTS.md

**Pull request closed without merging**